// @yext/pages/dist/scaffold/modules/templates.js
const formatModuleName = (moduleName) => {
  return moduleName.replace(/[-]+(\w)/g, (_, char) => char.toUpperCase()).replace(/^\w/, (char) => char.toUpperCase());
};
const moduleCode = (moduleName, useTailwind) => {
  const tailwind = useTailwind ? ` className="tailwind"` : ``;
  const formattedModuleName = formatModuleName(moduleName);
  return `import { Module, ModuleConfig, ModuleProps } from "@yext/pages/*";
import { AnalyticsProvider } from "@yext/pages-components";
import "./index.css";

const templateData: ModuleProps = {
  document: {
    businessId: "REPLACE_ME",
    siteId: "REPLACE_ME",
    __: {
      name: "${moduleName}",
      staticPage: true
    }
  }, 
  __meta: {
    mode: "production"
  }
}

export const config: ModuleConfig = {
  name: "${moduleName}"
}

const ${formattedModuleName}: Module = () => {
  return(
    <AnalyticsProvider 
      apiKey="REPLACE_ME"
      defaultCurrency="REPLACE_ME" 
      templateData={templateData}
      productionDomains={["REPLACE_ME"]}
    >
      <div${tailwind}>
        Module
      </div>
    </AnalyticsProvider>
  )
}

export default ${formattedModuleName};
`;
};
const postcssCode = () => {
  return `import path from "path";
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export default {
  plugins: {
    tailwindcss: { config: path.join(__dirname, 'tailwind.config.ts')},
    autoprefixer: {},
  },
};
`;
};
const indexCssCode = (useTailwind) => {
  return useTailwind ? `.tailwind {
  @tailwind base;
  @tailwind components;
  @tailwind utilities;  
}
` : ``;
};
const tailwindCode = (projectStructure) => {
  return `import type { Config } from 'tailwindcss';

export default {
  content: ["./${projectStructure.config.rootFolders.source}/**/*.{js,jsx,ts,tsx}"],
  theme: {
    extend: {},
  },
  plugins: [],
} satisfies Config
`;
};
export {
  formatModuleName,
  indexCssCode,
  moduleCode,
  postcssCode,
  tailwindCode
};
